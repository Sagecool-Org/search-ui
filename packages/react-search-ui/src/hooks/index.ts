export { useSearch } from "./use_search";
