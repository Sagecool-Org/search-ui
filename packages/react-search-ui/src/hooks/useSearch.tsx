import { SearchContextState } from "@elastic/search-ui";
import { useContext, useEffect, useState } from "react";

import SearchContext from "../SearchContext";

/**
 * React hook that provides access to Search UI state and actions
 *
 * @param mapContextToProps Optional function to select specific parts of the context
 * @returns Selected search context state and actions
 */
export function useSearch<T = SearchContextState>(
  mapContextToProps?: (context: SearchContextState) => T
): T {
  const context = useContext(SearchContext);

  if (!context) {
    throw new Error("useSearch must be used within a SearchProvider");
  }

  const [state, setState] = useState<T>(() =>
    mapContextToProps
      ? mapContextToProps({
          ...context.driver.getState(),
          ...context.driver.getActions()
        })
      : ({
          ...context.driver.getState(),
          ...context.driver.getActions()
        } as T)
  );

  useEffect(() => {
    const subscription = (newState: Partial<SearchContextState>) => {
      setState((prevState: T) => {
        const fullContext = {
          ...(prevState as any),
          ...newState
        };
        return mapContextToProps
          ? mapContextToProps(fullContext)
          : (fullContext as T);
      });
    };

    context.driver.subscribeToStateChanges(subscription);
    return () => {
      context.driver.unsubscribeToStateChanges(subscription);
    };
  }, [context.driver, mapContextToProps]);

  return state;
}

export default useSearch;
