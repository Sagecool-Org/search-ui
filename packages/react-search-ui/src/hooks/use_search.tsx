import { createContext, useContext } from "react";

import type { SearchContextState } from "@elastic/search-ui";

export const UseSearchContext = createContext<SearchContextState | undefined>(
  undefined
);

export const useSearch = () => {
  const context = useContext(UseSearchContext);
  if (context === undefined) {
    throw new Error("useSearch must be used within a SearchContextProvider");
  }
  return context;
};

export default useSearch;
