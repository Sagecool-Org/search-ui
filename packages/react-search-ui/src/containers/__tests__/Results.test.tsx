import React from "react";
import { render } from "@testing-library/react";
import ResultsContainer from "../Results";
import {
  ResultsViewProps,
  ResultViewProps
} from "@elastic/react-search-ui-views";
import { useSearch } from "../../hooks";

jest.mock("../../hooks", () => ({
  useSearch: jest.fn()
}));

const mockResults = [
  {
    id: {
      raw: "id",
      snippet: "<em>id</em>"
    },
    title: {
      raw: "title",
      snippet: "<em>title</em>"
    },
    url: {
      raw: "url",
      snippet: "<em>url</em>"
    }
  },
  {
    id: {
      raw: "id2",
      snippet: "<em>id</em>"
    },
    title: {
      raw: "title",
      snippet: "<em>title</em>"
    },
    url: {
      raw: "url",
      snippet: "<em>url</em>"
    }
  }
];

(useSearch as jest.Mock).mockReturnValue({ results: mockResults });

const params = {
  titleField: "title",
  urlField: "url"
};

it("supports a render prop", () => {
  const renderProp = ({ children }: ResultsViewProps) => {
    return <div>{children}</div>;
  };
  const { container } = render(
    <ResultsContainer {...params} view={renderProp} />
  );
  expect(container).toMatchSnapshot();
});

it("passes through props to individual Result items", () => {
  const resultView = ({ result }: ResultViewProps) => {
    return <li>{result.title.raw}</li>;
  };
  const { container } = render(
    <ResultsContainer
      {...params}
      resultView={resultView}
      shouldTrackClickThrough={true}
      clickThroughTags={["whatever"]}
    />
  );
  const resultItems = container.querySelectorAll("li");
  expect(resultItems).toHaveLength(2);
  resultItems.forEach((item) => {
    expect(item).toHaveTextContent("title");
  });
});

it("passes className through to the view", () => {
  let viewProps;
  const className = "test-class";
  render(
    <ResultsContainer
      {...params}
      className={className}
      view={(props) => {
        viewProps = props;
        return <div />;
      }}
    />
  );
  expect(viewProps.className).toEqual(className);
});

it("passes data-foo through to the view", () => {
  let viewProps;
  const data = "bar";
  render(
    <ResultsContainer
      {...params}
      data-foo={data}
      view={(props) => {
        viewProps = props;
        return <div />;
      }}
    />
  );
  expect(viewProps["data-foo"]).toEqual(data);
});
