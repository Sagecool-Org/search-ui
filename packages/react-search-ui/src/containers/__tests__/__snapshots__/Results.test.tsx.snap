// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`supports a render prop 1`] = `
<div>
  <div>
    <li
      class="sui-result"
    >
      <div
        class="sui-result__header"
      >
        <a
          class="sui-result__title sui-result__title-link"
          href="url"
          rel="noopener noreferrer"
          target="_blank"
        >
          <em>
            title
          </em>
        </a>
      </div>
      <div
        class="sui-result__body"
      >
        <ul
          class="sui-result__details"
        >
          <li>
            <span
              class="sui-result__key"
            >
              id
            </span>
             
            <span
              class="sui-result__value"
            >
              <em>
                id
              </em>
            </span>
          </li>
          <li>
            <span
              class="sui-result__key"
            >
              title
            </span>
             
            <span
              class="sui-result__value"
            >
              <em>
                title
              </em>
            </span>
          </li>
          <li>
            <span
              class="sui-result__key"
            >
              url
            </span>
             
            <span
              class="sui-result__value"
            >
              <em>
                url
              </em>
            </span>
          </li>
        </ul>
      </div>
    </li>
    <li
      class="sui-result"
    >
      <div
        class="sui-result__header"
      >
        <a
          class="sui-result__title sui-result__title-link"
          href="url"
          rel="noopener noreferrer"
          target="_blank"
        >
          <em>
            title
          </em>
        </a>
      </div>
      <div
        class="sui-result__body"
      >
        <ul
          class="sui-result__details"
        >
          <li>
            <span
              class="sui-result__key"
            >
              id
            </span>
             
            <span
              class="sui-result__value"
            >
              <em>
                id
              </em>
            </span>
          </li>
          <li>
            <span
              class="sui-result__key"
            >
              title
            </span>
             
            <span
              class="sui-result__value"
            >
              <em>
                title
              </em>
            </span>
          </li>
          <li>
            <span
              class="sui-result__key"
            >
              url
            </span>
             
            <span
              class="sui-result__value"
            >
              <em>
                url
              </em>
            </span>
          </li>
        </ul>
      </div>
    </li>
  </div>
</div>
`;
